pallets/stateful-storage/src/lib.rs

```diff
@@ -0,0 +1,746 @@
+			#[pallet::compact] target_hash: PageHash,
+			actions: BoundedVec<ItemAction, T::MaxItemizedActionsCount>,
+		) -> DispatchResult {
+			let provider_key = ensure_signed(origin)?;
+			let is_pruning = actions.iter().any(|a| matches!(a, ItemAction::Delete { .. }));
+			Self::check_actions(&actions)?;
+			Self::check_schema(schema_id, PayloadLocation::Itemized, false, is_pruning)?;
+			Self::check_grants(provider_key, state_owner_msa_id, schema_id)?;
+			Self::modify_itemized(state_owner_msa_id, schema_id, target_hash, actions)?;
+			Ok(())
+		}
+
+		/// Creates or updates an Paginated storage with new payload
+		#[pallet::call_index(1)]
+		#[pallet::weight(T::WeightInfo::upsert_page(payload.len() as u32))]
+		pub fn upsert_page(
+			origin: OriginFor<T>,
+			#[pallet::compact] state_owner_msa_id: MessageSourceId,
+			#[pallet::compact] schema_id: SchemaId,
+			#[pallet::compact] page_id: PageId,
+			#[pallet::compact] target_hash: PageHash,
+			payload: BoundedVec<u8, <T>::MaxPaginatedPageSizeBytes>,
+		) -> DispatchResult {
+			let provider_key = ensure_signed(origin)?;
+			ensure!(
+				page_id as u32 <= T::MaxPaginatedPageId::get(),
+				Error::<T>::PageIdExceedsMaxAllowed
+			);
+			Self::check_schema(schema_id, PayloadLocation::Paginated, false, false)?;
+			Self::check_grants(provider_key, state_owner_msa_id, schema_id)?;
+			Self::modify_paginated(
+				state_owner_msa_id,
+				schema_id,
+				page_id,
+				target_hash,
+				PaginatedPage::<T>::from(payload),
+			)?;
+			Ok(())
+		}
+
+		/// Deletes a Paginated storage
+		#[pallet::call_index(2)]
+		#[pallet::weight(T::WeightInfo::delete_page())]
+		pub fn delete_page(
+			origin: OriginFor<T>,
+			#[pallet::compact] state_owner_msa_id: MessageSourceId,
+			#[pallet::compact] schema_id: SchemaId,
+			#[pallet::compact] page_id: PageId,
+			#[pallet::compact] target_hash: PageHash,
+		) -> DispatchResult {
+			let provider_key = ensure_signed(origin)?;
+			ensure!(
+				page_id as u32 <= T::MaxPaginatedPageId::get(),
+				Error::<T>::PageIdExceedsMaxAllowed
+			);
+			Self::check_schema(schema_id, PayloadLocation::Paginated, false, true)?;
+			Self::check_grants(provider_key, state_owner_msa_id, schema_id)?;
+			Self::delete_paginated(state_owner_msa_id, schema_id, page_id, target_hash)?;
+			Ok(())
+		}
+
+		/// Applies the Add or Delete Actions on the requested Itemized page that requires signature
+		/// since the signature of delegator is checked there is no need for delegation validation
+		#[pallet::call_index(3)]
+		#[pallet::weight(T::WeightInfo::apply_item_actions_with_signature( payload.actions.len() as u32 ,
+			payload.actions.iter().fold(0, |acc, a| acc + match a {
+			ItemAction::Add { data } => data.len() as u32,
+			_ => 0,
+			})
+		))]
+		pub fn apply_item_actions_with_signature(
+			origin: OriginFor<T>,
+			delegator_key: T::AccountId,
+			proof: MultiSignature,
+			payload: ItemizedSignaturePayload<T>,
+		) -> DispatchResult {
+			ensure_signed(origin)?;
```

Comment:
We don't need to call `check_grants`, since we verify that the signer of the payload is the owner of the msa_id being acted upon.

Further question, though: do we want to additionally support a use case where a _delegate_ can issue a signed payload? Right now it looks like we don't.

Reply:
One of the use-cases of these extrinisics is to allow **on-time delegation**. 

> Further question, though: do we want to additionally support a use case where a delegate can issue a signed payload? Right now it looks like we don't.

I don't think we need that at the moment but we can always add later if that scenario became important.

Reply:
More specifically I was asking if it's intended to allow these types of actions to be delegated to an MSA that is not a registered Provider.  It was clear that the signature itself and the payload are validated.

Reply:
Okay, just making sure that was intentional.